### ISSUE TITLE
ROS1 MAVROS Decode Failure: Decoded data is shorter than predicted value length

### ISSUE THREAD

**Issue by your-sudden-death** [Carter12s/roslibrust#148]
Hi everyone,
when trying to decode a `/mavros/global_position/global` (type `NavSatFix`) message with this crate (using ROS1 natively),
the decoder fails with
```
2024-01-28T14:23:11.616115Z DEBUG reqwest::connect: starting new connection: http://localhost:11311/
2024-01-28T14:23:11.616335Z DEBUG hyper::client::connect::dns: resolving host="localhost"
2024-01-28T14:23:11.616598Z DEBUG hyper::client::connect::http: connecting to 127.0.0.1:11311
2024-01-28T14:23:11.616710Z DEBUG hyper::client::connect::http: connected to 127.0.0.1:11311
2024-01-28T14:23:11.616991Z DEBUG hyper::proto::h1::io: flushed 254 bytes
2024-01-28T14:23:11.617371Z DEBUG hyper::proto::h1::io: parsed 4 headers
2024-01-28T14:23:11.617387Z DEBUG hyper::proto::h1::conn: incoming body is content-length (258 bytes)
2024-01-28T14:23:11.617418Z DEBUG hyper::proto::h1::conn: incoming body completed
2024-01-28T14:23:11.617477Z DEBUG hyper::client::pool: pooling idle connection for ("http", localhost:11311)
2024-01-28T14:23:11.617823Z DEBUG hyper::client::pool: reuse idle connection for ("http", localhost:11311)
2024-01-28T14:23:11.617908Z DEBUG hyper::proto::h1::io: flushed 466 bytes
2024-01-28T14:23:11.618458Z DEBUG hyper::proto::h1::io: parsed 4 headers
2024-01-28T14:23:11.618465Z DEBUG hyper::proto::h1::conn: incoming body is content-length (277 bytes)
2024-01-28T14:23:11.618481Z DEBUG hyper::proto::h1::conn: incoming body completed
2024-01-28T14:23:11.618515Z DEBUG hyper::client::pool: pooling idle connection for ("http", localhost:11311)
2024-01-28T14:23:11.618731Z DEBUG hyper::client::pool: reuse idle connection for ("http", localhost:11311)
2024-01-28T14:23:11.618792Z DEBUG hyper::proto::h1::io: flushed 483 bytes
2024-01-28T14:23:11.619146Z DEBUG hyper::proto::h1::io: parsed 4 headers
2024-01-28T14:23:11.619154Z DEBUG hyper::proto::h1::conn: incoming body is content-length (349 bytes)
2024-01-28T14:23:11.619166Z DEBUG hyper::proto::h1::conn: incoming body completed
2024-01-28T14:23:11.619202Z DEBUG hyper::client::pool: pooling idle connection for ("http", localhost:11311)
2024-01-28T14:23:11.623457Z DEBUG reqwest::connect: starting new connection: http://3a8ab27ba290:42315/
2024-01-28T14:23:11.623513Z DEBUG hyper::client::connect::dns: resolving host="3a8ab27ba290"
2024-01-28T14:23:11.623594Z DEBUG hyper::client::connect::http: connecting to 172.21.0.2:42315
2024-01-28T14:23:11.623710Z DEBUG hyper::client::connect::http: connected to 172.21.0.2:42315
2024-01-28T14:23:11.623848Z DEBUG hyper::proto::h1::io: flushed 479 bytes
2024-01-28T14:23:11.623936Z DEBUG hyper::proto::h1::io: parsed 3 headers
2024-01-28T14:23:11.623944Z DEBUG hyper::proto::h1::conn: incoming body is content-length (298 bytes)
2024-01-28T14:23:11.623961Z DEBUG hyper::proto::h1::conn: incoming body completed
2024-01-28T14:23:11.623997Z DEBUG hyper::client::pool: pooling idle connection for ("http", 3a8ab27ba290:42315)
2024-01-28T14:23:11.624093Z DEBUG roslibrust::ros1::subscriber: Got a TCPROS publisher endpoint at 3a8ab27ba290:37625
2024-01-28T14:23:11.624873Z DEBUG roslibrust::ros1::subscriber: Established connection with publisher for /mavros/global_position/global
2024-01-28T14:23:11.635978Z DEBUG roslibrust::ros1::subscriber: Read 129 bytes from the publisher connection
2024-01-28T14:23:11.636128Z  INFO testsuite: Error(Underflow, State { next_error: None, backtrace: None })
2024-01-28T14:23:11.656456Z DEBUG roslibrust::ros1::subscriber: Read 129 bytes from the publisher connection
2024-01-28T14:22:22.735653Z  INFO testsuite: Decoded data is shorter than predicted value length
```

This is on ROS1 Noetic, package versions below (latest crates.io release of roslibrust / codegen)
Code in Main:
```rust
  tracing_subscriber::fmt()
      .with_max_level(Level::DEBUG)
      .init();

  let nh = NodeHandle::new("http://localhost:11311", "/testsuite")
      .await
      .unwrap();

  let mut get_position_global = nh
      .subscribe::<sensor_msgs::NavSatFix>("/mavros/global_position/global", 16)
      .await
      .unwrap();

  loop {
      match get_position_global.next().await {
          Ok(_) => (),
          Err(e) => info!("{:?}", e),
      }
  }
```

The common_msgs are on commit `20a833b56f9d7fd39655b8491a2ec1226d2639b3`
Rostopic Echo shows:
```
---
header:
  seq: 47246
  stamp:
    secs: 1425
    nsecs: 240000000
  frame_id: "base_link"
status:
  status: 0
  service: 1
latitude: 47.3977419
longitude: 8.545594
altitude: 535.3779155639569
position_covariance: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
position_covariance_type: 2
---
```

I suspect it has something to do with the time field of the Header, Messages from `std_msgs` work (even when emitted by MAVROS)

**Comment by Carter12s** [comment]
Thank you for a truly excellent bug report with wonderful detail. Will dig in soon.

**Comment by Carter12s** [comment]
So according to the git history here:
https://github.com/ros/common_msgs/blob/noetic-devel/sensor_msgs/msg/NavSatFix.msg

NavSatFix hasn't had any changes to its definition in the last 12 years, so this shouldn't be caused by a schema change.
The ROS2 definition is "identical": https://docs.ros2.org/latest/api/sensor_msgs/msg/NavSatFix.html
But if it was being used, would cause deserialization to fail because ROS2 defines the timestamp differently. That may be the problem here, @your-sudden-death can you add the code you used for message generation of the rust messages and confirm the type you are using in `.subscribe<sensor_msgs::NavSatFix>` is generated from the ROS1 definition of the message?

---

NEVERMIND that might also be a problem, but isn't the only problem...

This issue replicates locally for me on `master`, simply by sending the message with `rostopic pub`. Digging in more...

---

Okay I've tracked the failure directly to our call to `serde_rosmsg::from_slice` inside of Subscriber::next. So it looks like this failure is internal to the serde_rosmsg crate. Will continue to dig, but fix may come slower, as we'll likely need to open an issue over there to get this resolved.

**Comment by Carter12s** [comment]
Okay I've actually root caused the problem now.

The NavSatFix message has a fixed size array in it (rarely used in ROS1).

roslibrust is representing the array (even thou it is fix sized) as `Vec<f64>` instead of `[f64; 9]` this causes serde_rosmsg to fail to deserialize as fixed sized arrays actually have a different byte representation over the wire.

To fix this we actually need to upgrade the codegen for roslibrust to generate fix sized arrays as arrays instead of vecs. We got away with this previously because in the JSON format rosbridge uses, the two are identical over the wire.

**Comment by Carter12s** [comment]
Okay so I started on a implementation to fix this today and ran into an immediate snag...

Our crate and serde_rosmsg that we depend on for ROS1 serialization rely on certain trait definitions existing on types (Default, Debug, Serialize, etc....). Rust's type system currently doesn't have a way for a trait to be generic over a fixed size array of any size. There is an RFC feature to enable this https://rust-lang.github.io/rfcs/2000-const-generics.html and the feature has been available in beta since Rust 1.51, but we don't have great access to it yet.

Until this features lands authors of traits have to manually (usually with macros) generate and implementation of that trait for each possible size of a fixed sized array. Typically authors follow the standard libraries decision here, and impl all sizes from 1 to 32.

Unfortunately in the "standard messages" for both ROS1 and ROS2 there is "TwistWithCovariance" message that includes a 36 member fixed size array, and this type is currently breaking a ton of our tests because code generated by codegen can't compile:

Example error:
```
error[E0277]: the trait bound `[f64; 36]: Serialize` is not satisfied
    --> roslibrust_test/src/ros2.rs:687:9
     |
687  |         :: serde :: Serialize,
     |         ^^^^^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `[f64; 36]`
...
695  |         pub r#covariance: [f64; 36],
     |         --- required by a bound introduced by this call
     |
     = help: the following other types implement trait `Serialize`:
               [T; 0]
               [T; 1]
               [T; 2]
               [T; 3]
               [T; 4]
               [T; 5]
               [T; 6]
               [T; 7]
             and 26 others
note: required by a bound in `ros1::actionlib_msgs::_::_serde::ser::SerializeStruct::serialize_field`
```

Will require some thought on how we want to handle this limitation for now...

**Comment by Carter12s** [comment]
I believe I've found an effective work around relying on the [serde-big-array](https://crates.io/crates/serde-big-array).
Heading for testing / review on the branch.
I think this is a big enough fix that we'll cut v0.9 of this lib once this fix lands.
